LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/UpstreamDataFrame.cs

```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoRaWan.NetworkServer.BasicsStation
+{
+    public class UpstreamDataFrame
+    {
+        public UpstreamDataFrame(DevAddr devAddr, ushort frameCounter, string fRMPayload, Mic mic)
+        {
+            DevAddr = devAddr;
+            FrameCounter = frameCounter;
+            FRMPayload = fRMPayload;
+            Mic = mic;
+        }
+
+        public virtual DevAddr DevAddr { get; }
```

Comment:
These properties were marked as virtual so that we could setup a mock on them. Is there a way to avoid this? It's also marked as an issue from CodeQL scanning.

Reply:
An interface would work, I don't know other way. That's an interesting question

Reply:
Would work yes but not sure if an interface is a good choice for a DTO class like this.

Reply:
If it's a pure data class that has no behavioural component then it doesn't need to be mocked at all therefore don't need an interface or make them virtual.

Reply:
It is a pure data class.

The `virtual` was added for the test method `ConcentratorDeduplicationTest.CachedEntries_Should_Expire`: do you have a better idea how we can test this without making the properties virtual? Or do you think we should drop the test alltogether (imo, I think it adds value)

Reply:
Do you mean `CacheKey_Should_Consider_All_Required_Fields`?

Reply:
Sorry 🤦‍♂️ yes, you are right

Reply:
If you did indeed mean `CacheKey_Should_Consider_All_Required_Fields` then I think the test should test the generated cache key instead of testing that the properties of the object were simply accessed. Access to properties does not guarantee that they are part of the key. Consequently, you can make the properties non-virtual.

Reply:
Agree that it's not proof that they are part of the key. [Changed the test to test for an expected result instead and made the properties non-virtual again](47dbfc79f9c0a74442924160e7b9178d75616bc6 ).